operator/src/main/java/org/keycloak/operator/v2alpha1/WatchedSecretsStore.java

```diff
@@ -0,0 +1,237 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.keycloak.operator.v2alpha1;
+
+import io.fabric8.kubernetes.api.model.HasMetadata;
+import io.fabric8.kubernetes.api.model.Secret;
+import io.fabric8.kubernetes.api.model.SecretBuilder;
+import io.fabric8.kubernetes.client.KubernetesClient;
+import io.fabric8.kubernetes.client.informers.SharedIndexInformer;
+import io.fabric8.kubernetes.client.utils.Serialization;
+import io.javaoperatorsdk.operator.processing.event.ResourceID;
+import io.javaoperatorsdk.operator.processing.event.source.EventSource;
+import io.javaoperatorsdk.operator.processing.event.source.informer.InformerEventSource;
+import io.javaoperatorsdk.operator.processing.event.source.informer.Mappers;
+import io.quarkus.logging.Log;
+import org.keycloak.operator.Constants;
+import org.keycloak.operator.OperatorManagedResource;
+import org.keycloak.operator.v2alpha1.crds.Keycloak;
+
+import java.math.BigInteger;
+import java.nio.charset.StandardCharsets;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.Base64;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * Represents a version store of Secrets that are watched by a CR but is not owned by it. E.g. Secrets with
+ * credentials provided by user.
+ *
+ * It is backed by a Secret which holds a list of watched Secrets together with their last observed version. It marks
+ * all the watched Secrets with a label indicating which CRs are watching that resource.
+ *
+ * @author Vaclav Muzikar <vmuzikar@redhat.com>
+ */
+public class WatchedSecretsStore extends OperatorManagedResource {
+    public static final String TYPE = "secrets-store";
+    public static final String TYPE_WATCHED = "watched-secret";
+    public static final String STORE_SUFFIX = "-" + TYPE;
+
+    private final Secret existingStore; // a Secret to store the last observed versions
+
+    // key is name of the secret
+    private final Map<String, String> lastObservedVersions;
+    private final Map<String, String> currentVersions;
+    private final Set<Secret> currentSecrets;
+
+    public WatchedSecretsStore(Set<String> desiredWatchedSecretsNames, KubernetesClient client, Keycloak kc) {
+        super(client, kc);
+        existingStore = fetchExistingStore();
+        lastObservedVersions = getNewLastObservedVersions();
+        currentSecrets = fetchCurrentSecrets(desiredWatchedSecretsNames);
+        currentVersions = getNewCurrentVersions();
+    }
+
+    /**
+     * @return true if any of the watched Secrets was changed, false otherwise (incl. if it's a newly watched Secret)
+     */
+    public boolean changesDetected() {
+        return currentVersions.entrySet().stream().anyMatch(e -> {
```

Comment:
I don't think so. What would it tell us? That we've added or removed a secret, neither of which requires an explicit restart because it's done implicitly by K8s.

Reply:
> neither of which requires an explicit restart because it's done implicitly by K8s.

can you elaborate?
k8s doesn't restart pods in case of secrets changes.

Reply:
If `currentVersions.size() != lastObservedVersions.size()` that means that the user either used a Secret in the config that was not used before, or the opposite – removed a Secret from the config. In any case that means the config was changed (not the Secret itself). If the config was changed, the Deployment restart is implicitly done by K8s as the env vars list was changes.